### ISSUE TITLE
[Bug]: Generation fails with spec issues

### ISSUE THREAD

**Issue by Haarolean** [PaddleHQ/paddle-openapi#5]
### What happened?

Some parts of the contract contain invalid spec syntax, like this one: https://github.com/PaddleHQ/paddle-openapi/blob/main/v1/openapi.yaml#L4170

```
There were issues with the specification. The option can be disabled via validateSpec (Maven/Gradle) or --skip-validate-spec (CLI).
 | Error count: 4, Warning count: 42
Errors: 
	-attribute paths.'/reports'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}/credit-balances'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/discounts'(get).parameters.[['unknown']].in is missing
Warnings: 
	-attribute paths.'/reports'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}/credit-balances'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/discounts'(get).parameters.[['unknown']].in is missing
```

### Steps to reproduce

1. build via openapi-generator
2. profit

### What did you expect to happen?

to get a bunch of generated classes

### Logs

```shell
org.openapitools.codegen.SpecValidationException: There were issues with the specification. The option can be disabled via validateSpec (Maven/Gradle) or --skip-validate-spec (CLI).
 | Error count: 4, Warning count: 42
Errors: 
	-attribute paths.'/reports'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}/credit-balances'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/discounts'(get).parameters.[['unknown']].in is missing
Warnings: 
	-attribute paths.'/reports'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}/credit-balances'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/customers/{customer_id}'(get).parameters.[['unknown']].in is missing
	-attribute paths.'/discounts'(get).parameters.[['unknown']].in is missing
```

**Comment by vijayasingam-paddle** [comment]
Hi @Haarolean,
Thank you for raising this issue to us. 
I am sorry about this issue. Upon analysis, I can see that one of our script is not working well and are leaving empty objects `{}` behind. 

I will let you know once this issue is fixed. 

Thank you.

**Comment by dvrfluxchat** [comment]
hey @vijayasingam-paddle , any idea when this will be done? I am using cloudflare pages to deploy my nextjs app but that is not supported because of some edge runtime issues. I want to create a typesafe client but all the codegen tools are breaking with this openapi spec. would really appreciate any timeline here

**Comment by vijayasingam-paddle** [comment]
Hello 👋 
Sorry for not providing an update to this thread for a while. 

We did fix the issue with empty objects(`{}`) but then realized that the `openapi-generator` was still not happy due to `x-enum-descriptions` and some other issues with `oneOf` 

We are trying to find the best way to fix the other issues we encountered.
I will let you know once I can successfully generate an SDK.


@dvrfluxchat,
I'm sorry that you had to generate your types manually due to an issue with our Node SDK. 
Both issues should be fixed in the next few weeks. However, I recommend using our Node SDK instead of generating types as we can't guarantee how accurate they will be.

Thank you.

**Comment by dvrfluxchat** [comment]
Node SDK does not work in edge environment because of the node:crypto module being used in few places, for now i forked the sdk and removed the dependencies and it works fine in edge.

**Comment by Haarolean** [comment]
Are there any updates here?

**Comment by SaadArdati** [comment]
This issue is still pending...

**Comment by Stephan3555** [comment]
I ran into the same issue — it seems the OpenAPI schema isn't compatible with OpenAPI Generator.
Any updates on this? @vijayasingam-paddle  @heymcgovern 

A fix would be greatly appreciated, as it would significantly simplify the integration with Paddle.

**Comment by heymcgovern** [comment]
Hey all 👋 Thanks for your patience with us on this one. I spent some time digging into this and it ultimately looks like this is a problem with `openapi-generator` and OpenAPI 3.1 support, which they note is in beta.

I'll walk through my investigation in case it helps 👇

Before we start, I spotted a small bug in our spec where `status_report_query` has `default: active`, but this isn't a valid enum. I have a fix for this in review and it should be live at some point next week.

First, I ran the script through [`openapi-spec-validator`](https://pypi.org/project/openapi-spec-validator/) and it comes back as valid OAS 3.1:

```
openapi-spec-validator openapi.yaml
openapi.yaml: OK
```

Second, I created a script to strip `x-enum-descriptions` from the spec. We use these to describe enums, but they're structured differently to the way `openapi-generator` expects, so I removed them. Here's the script I used, if you're interested: https://gist.github.com/heymcgovern/d9347a5ae5966dad6ced27409a60538d

Next, I ran through `openapi-generator`:

```
npx @openapitools/openapi-generator-cli generate -i openapi_modified.yaml -g java -o ./tmp/test/
```

There are a bunch of non-fatal warnings and errors like this:

```
[main] WARN  o.o.codegen.utils.ModelUtils - Failed to get the schema name: null
[main] ERROR o.o.codegen.OpenAPINormalizer - Type {} not yet supported in openapi-normalizer to process OpenAPI 3.1 spec with multiple types.
```

My best guess is that this is around `type` arrays and `null` types introduced in OAS 3.1. We use these extensively in our spec, and it's one of the reasons that [adding support for 3.0 isn't feasible](https://github.com/PaddleHQ/paddle-openapi/issues/3). Here's a quick comparison of OAS 3.0 vs 3.1:

```
# OpenAPI 3.0
example_field:
  type: string
  nullable: true

# OpenAPI 3.1
example_field:
  type:
    - "string"
    - "null" 
```

It looks like this issue doesn't block generation, but might result in incorrect types or methods being generated. 

Following this, there's a fatal error:

```
Could not process model 'simulation'.Please make sure that your schema is correct!
```

After some trial and error, it looks like the generator has trouble with the `simulation` and `simulation_scenario_create_config` schemas. Removing the `oneOf` and `discriminator` elements from these schemas results in successful generation, suggesting there's an issue with polymorphism.

The structure for these is valid and accurately describes the way the API behaves. We use the spec without issue in tooling and workflows on our side for testing and validation, SDK generation, and our developer docs, so it seems like the issue might be with `openapi-generator`. OpenAPI 3.1 support is in beta, so it's likely an outstanding issue.

### What you can do

If you want to proceed with generation, you can:

1. Strip `x-enum-descriptions` from the spec.
2. Strip simulations schemas and endpoints from the spec. Simulations won't be available in your generated SDKs, but everything else should be present.
3. Review the generated SDKs carefully and check them against [our docs](https://developer.paddle.com/api-reference/overview) or the other SDKs to make sure they're correct. I'd recommend paying particular attention to types given the `OpenAPINormalizer` error.

### What we've done

* ✅ We previously fixed an issue with our tooling where it output empty objects for parameters, like `- {}`. This is in the latest version of the spec.
* ✅ We have a quick fix for `status_report_query` in the works. This didn't stop generation in my testing, but definitely nice to have for accuracy.
* ⏭️ We're going to review how we document enum descriptions in a future version of our API. We're not going to strip these or change how we do them in this version, since users may rely on them in workflows they've built.
* ⏭️ I created a ticket to review our simulations schema structure. This is our only `discriminator` pattern, so it's possible that we could simplify it. Changes here require careful testing with existing workflows, so this is in our backlog rather than immediate roadmap.

Since it looks like this is related to OpenAPI 3.1 support in `openapi-generator`, there's not much more we can do from our side at the moment. Hope this investigation helps you troubleshoot and find a path forward. Let me know if you need anything else here.

-- Michael

**Comment by heymcgovern** [comment]
Hey everyone 👋 Just a quick one to let you know that I updated the spec with the fix to remove `default: active` from the `status_report_query` schema.

I'm going to close this issue because it looks like the issue is outside of our control, but please pop back if there's anything else we can do to help.

**Comment by nob13** [comment]
@heymcgovern can you provide us with a script or a derived openapi.yaml which we can use in the openapi-generator?
